server-common/src/main/java/org/apache/kafka/timeline/SnapshottableHashTable.java

```diff
@@ -113,7 +113,7 @@ public void mergeFrom(long epoch, Delta source) {
             HashTier<T> other = (HashTier<T>) source;
             // As an optimization, the deltaTable might not exist for a new key
             // as there is no previous value
-            if (other.deltaTable != null) {
+            if (deltaTable != null && other.deltaTable != null) {
```

Comment:
Hmm, if our deltaTable doesn't exist here, we should create it like this, right?
```
            if (other.deltaTable != null) {
                List<T> list = new ArrayList<>();
                Object[] otherElements = other.deltaTable.baseElements();
                for (int slot = 0; slot < otherElements.length; slot++) {
                    BaseHashTable.unpackSlot(list, otherElements, slot);
                    for (T element : list) {
                        // When merging in a later hash tier, we want to keep only the elements
                        // that were present at our epoch.
                        if (element.startEpoch() <= epoch) {
                            if (deltaTable == null) {
                                deltaTable = new BaseHashTable<>(1);
                            }
                            deltaTable.baseAddOrReplace(element);
                        }
                    }
                }
            }
```

Otherwise we're not merging in all the stuff we should be from the later snapshot

Reply:
Ah, you are right! We need to copy the other deltatable to `this`. Will update it later.